scripts/globals/fishing_contest.lua

```diff
@@ -0,0 +1,417 @@
+        player:messageSpecial(messageId, fishid)
+        return false
+    end
+
+    -- give items to player
+    if player:addItem(params) then
+        player:messageSpecial(ID.text.ITEM_OBTAINED, fishid)
+    else
+        player:messageSpecial(ID.text.ITEM_CANNOT_BE_OBTAINED, fishid)
+        return false
+    end
+
+    return true
+end
+
+local function getAwardInteger(awardTable)
+    return bit.lshift(awardTable[4], 24) + bit.lshift(awardTable[3], 16) + bit.lshift(awardTable[2], 8) + awardTable[1]
+end
+
+-- Determine if the player has a reward pending in the current contest
+-- This will not permit rewarding for ongoing contests until it is ACTUALLY at the presenting stage.
+local function isRewardAvailable(player)
+    local awards = player:getContestRewardStatus()
+
+    if awards ~= nil then
+        if
+            awards['rank'] > 0 and
+            awards['rank'] <= 20
+        then
+            return 1
+        end
+    end
+
+    return 0
+end
+
+-- This returns the data necessary to send the packet with the 'time remaining until stage change' data to the player
+local function getTimeRemaining(changeTime)
+    local timeTable = {}
+    local currentTime = os.time()
+    if changeTime > currentTime then
+        local timeDelta = changeTime - currentTime
+        timeTable =
+        {
+            ['days']    = math.floor(timeDelta / (60 * 60 * 24)),
+            ['hours']   = math.floor((timeDelta / (60 * 60)) % 24),
+            ['minutes'] = math.floor((timeDelta / 60) % 60),
+        }
+    else
+        timeTable =
+        {
+            ['days']    = 0,
+            ['hours']   = 0,
+            ['minutes'] = 0,
+        }
+    end
+
+    return timeTable
+end
+
+-- Provides the score of the submission
+local function scoreFish(length, weight, criteria)
+    local score = 0
+    if criteria == xi.fishingcontest.criteria.SIZE then
+        score = length
+    elseif criteria == xi.fishingcontest.criteria.WEIGHT then
+        score = weight
+    else
+        score = length + weight
+    end
+
+    return score
+end
+
+-- Determines the rewards to be granted based on rank and share
+local function getPlayerReward(player)
+    local status = player:getContestRewardStatus()
+
+    if status ~= nil then
+        local rank  = status['rank']
+        local share = status['share'] or 1
+        if
+            rank > 0 and
+            rank <= 20
+        then
+            local reward = xi.fishingcontest.reward[rank]
+            if reward.gil then
+                reward.gil = reward.gil / share
+            end
+
+            return reward
+        end
+    end
+
+    return nil
+end
+
+-----------------------------------
+-- HOOKED FUNCTIONS
+-----------------------------------
+xi.fishingcontest.selectContestFish = function()
+    local fishList = xi.fishingcontest.fish
+    return xi.fishingcontest.fish[math.random(#fishList)]['id']
```

Comment:
We have utils to wrap this so you can do utils.randomEntry(fishList)['id'], I think